packages/flutter/lib/src/widgets/interactive_viewer.dart

```diff
@@ -0,0 +1,1136 @@
+        return Listener(
+          onPointerSignal: _receivedPointerSignal,
+          child: GestureDetector(
+            behavior: HitTestBehavior.opaque, // Necessary when translating off screen
+            onScaleEnd: _onScaleEnd,
+            onScaleStart: _onScaleStart,
+            onScaleUpdate: _onScaleUpdate,
+            child: child,
+          ),
+        );
+      },
+    );
+  }
+}
+
+/// A thin wrapper on [ValueNotifier] whose value is a [Matrix4] representing a
+/// transformation.
+///
+/// The [value] defaults to the identity matrix, which corresponds to no
+/// transformation.
+///
+/// See also:
+///
+///  * [InteractiveViewer.transformationController] for detailed documentation
+///    on how to use TransformationController with [InteractiveViewer].
+class TransformationController extends ValueNotifier<Matrix4> {
+  /// Create an instance of [TransformationController].
+  ///
+  /// The [value] defaults to the identity matrix, which corresponds to no
+  /// transformation.
+  TransformationController([Matrix4 value]) : super(value ?? Matrix4.identity());
+
+  /// Return the scene point at the given viewport point.
+  ///
+  /// A viewport point is relative to the parent while a scene point is relative
+  /// to the child, regardless of transformation. Calling toScene with a
+  /// viewport point essentially returns the scene coordinate that lies
+  /// underneath the viewport point given the transform.
+  ///
+  /// The viewport transforms as the inverse of the child (i.e. moving the child
+  /// left is equivalent to moving the viewport right).
+  ///
+  /// This method is often useful when determining where an event on the parent
+  /// occurs on the child. This example shows how to determine where a tap on
+  /// the parent occurred on the child.
+  ///
+  /// ```dart
+  /// @override
+  /// void build(BuildContext context) {
+  ///   return GestureDetector(
+  ///     onTapUp: (TapUpDetails details) {
+  ///       _childWasTappedAt = _transformationController.toScene(
+  ///         details.localPosition,
+  ///       );
+  ///     },
+  ///     child: InteractiveViewer(
+  ///       transformationController: _transformationController,
+  ///       child: child,
+  ///     ),
+  ///   );
+  /// }
+  /// ```
+  Offset toScene(Offset viewportPoint) {
+    // On viewportPoint, perform the inverse transformation of the scene to get
+    // where the point would be in the scene before the transformation.
+    final Matrix4 inverseMatrix = Matrix4.inverted(value);
+    final Vector3 untransformed = inverseMatrix.transform3(Vector3(
+      viewportPoint.dx,
+      viewportPoint.dy,
+      0,
+    ));
+    return Offset(untransformed.x, untransformed.y);
+  }
+}
+
+// A classification of relevant user gestures. Each contiguous user gesture is
+// represented by exactly one _GestureType.
+enum _GestureType {
+  translate,
+  scale,
+  rotate,
+}
+
+/// Returns the closest point to the given point on the given line segment.
```

Comment:
Wondering if some (all?) of these should be static members of the class above to avoid putting them into the public namespace (even when tagged as visibleForTesting they will show up in our docs, I believe).

Reply:
I moved them to be static methods on InteractiveViewer with `@visibleForTesting`.